gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/Client.java

```diff
@@ -742,19 +750,42 @@ protected Connection chooseConnection(final RequestMessage msg) throws TimeoutEx
          */
         @Override
         protected void initializeImplementation() {
-            // chooses an available host at random
-            final List<Host> hosts = cluster.allHosts()
+            // chooses an available host at random from available hosts
+            List<Host> hosts = cluster.allHosts()
                     .stream().filter(Host::isAvailable).collect(Collectors.toList());
-            if (hosts.isEmpty()) throw new IllegalStateException("No available host in the cluster");
+            // if there is no available hosts, check for all hosts
+            if (hosts.isEmpty()) {
+                hosts = new ArrayList<>(cluster.allHosts());
+                if (hosts.isEmpty()) {
+                    throw new IllegalStateException("No available host in the cluster");
+                }
+            }
             Collections.shuffle(hosts);
             final Host host = hosts.get(0);
 
+            // only mark host as available if we can initialize connection pool successfully
+            final CompletableFuture<Void> completableFuture =
+                    CompletableFuture.runAsync(() -> initializeConnectionSetupForHost.accept(host), cluster.executor());
+            try {
+                completableFuture.join();
+                host.makeAvailable();
+            } catch (CompletionException ex) {
+                Throwable result = ex;
+                Throwable cause = ex.getCause();
+                result = cause == null ? result : cause;
+                logger.error("", result);
+            }
+        }
+
+        private Consumer<Host> initializeConnectionSetupForHost = host -> {
             try {
+                // hosts that don't initialize connection pools will come up as a dead host
                 connectionPool = new ConnectionPool(host, this, Optional.of(1), Optional.of(1));
             } catch (RuntimeException ex) {
-                logger.error("Could not initialize client for {}", host, ex);
+                final String errMsg = "Could not initialize client for " + host;
+                throw new RuntimeException(errMsg, ex);
```

Comment:
Is this supposed to be a RuntimeException, or should it be a CompletionException, based on your block above that joins this? I haven't used CompletableFuture try/catch much in my own code so I am not sure.

Reply:
I'm not sure either, this is a direct copy from their original code, so I'm not sure if we should change it?

Reply:
Best to educate ourselves if neither of us know. Check this out https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#join--

According to this, if the future completed with an unchecked exception (such as our RuntimeException), it will wrap that exception with a CompletionException before propagating it. Therefore this is safe.

Also, it's best practice to use String.format in the error message above, RuntimeException(String.format("Could not initialize client for %s.", host), ex);

Reply:
Sounds good